include/zephyr/rtio/rtio_spsc.h

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#ifndef ZEPHYR_RTIO_SPSC_H_
+#define ZEPHYR_RTIO_SPSC_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+
+/**
+ * @brief RTIO SPSC API
+ * @defgroup rtio_spsc RTIO SPSC API
+ * @ingroup rtio
+ * @{
+ */
+
+/**
+ * @file rtio_spsc.h
+ *
+ * @brief A lock-free and type safe power of 2 fixed sized SPSC Queue using a
+ * ringbuffer and atomics to ensure coherency.
+ *
+ * This SPSC implementation works on an array which wraps using a power of two
+ * size and uses a bit mask to perform a modulus. Atomics are used to allow
+ * single-producer single-consumer safe semantics without locks. Elements are
+ * expected to be of a fixed size. The API is type safe as the underlying buffer
+ * is typed and all usage is done through macros.
+ *
+ * A spsc may be declared on a stack or statically and work as intended so long
+ * as its lifetime outlives any usage. Static declarations should be the
+ * preferred method. It is meant to be a shared object between two execution
+ * contexts (ISR and a thread for example)
+ *
+ * The SPSC is safe to produce or consume in an ISR with O(1) push/pull.
+ *
+ * The SPSC is *not* safe to produce in multiple execution contexts or consume in
+ * multiple execution contexts without additional work.
+ *
+ * Safe usage would be, where A and B are unique execution contexts:
+ * 1. ISR A producing and a Thread B consuming.
+ * 2. Thread A producing and ISR B consuming.
+ * 3. Thread A producing and Thread B consuming.
+ */
+
+/**
+ * @private
+ * @brief Common spsc attributes
+ *
+ * @warning Not to be manipulated without the macros!
+ */
+struct rtio_spsc {
+	/* private value only the producer thread should mutate */
+	unsigned long acquire;
+
+	/* private value only the consumer thread should mutate */
+	unsigned long consume;
+
+	/* producer mutable, consumer readable */
+	atomic_t in;
+
+	/* consumer mutable, producer readable */
+	atomic_t out;
+
+	/* mask used to automatically wrap values */
+	const unsigned long mask;
+};
+
+/**
+ * @brief Statically initialize a rtio_spsc
```

Comment:
if this will be read as "initialize a Real Time IO" then 'a' is correct, otherwise if rtio_spsc will be read as R T I O S P S C, then "initialize an rtio_spsc"